core/indexer/dataDispatcher.go

```diff
@@ -54,18 +61,55 @@ func (d *dataDispatcher) startWorker(ctx context.Context) {
 			log.Debug("dispatcher's go routine is stopping...")
 			return
 		case wi := <-d.chanWorkItems:
-			d.doWork(wi)
+			d.writeMutex.Lock()
+			d.doWork(wi, 0)
+			d.writeMutex.Unlock()
 		}
 	}
 }
 
 // Close will close the endless running go routine
 func (d *dataDispatcher) Close() error {
+	if d.wasClosed.IsSet() {
+		return nil
+	}
+
+	start := time.Now()
+	d.wasClosed.Set()
+
 	if d.cancelFunc != nil {
 		d.cancelFunc()
 	}
 
-	return nil
+	d.writeMutex.Lock()
```

Comment:
we can have a deadlock here: suppose the mutex was captured by L64, this will never Close